### ISSUE TITLE
undefined page references in l3draw-code

### ISSUE THREAD

**Issue by mbertucci47** [latex3/latex3#1397]
This is the same as https://github.com/latex3/latex3/issues/1345 but for l3draw-code, namely undefined page references in the sentences "These functions are documented on page ??". Here all of the non-internal functions produce the undefined reference.

**Comment by muzimuzhi** [comment]
I find `l3draw` is the only `l3experimental` package which, apart from a `<pkgname>.pdf`, provides an extra `<pkgname>-code.pdf`, and it's also the only package which has its documented source (the implementation part) typeset.

The problem in `l3draw-code.pdf` is that it only contains implementation part (unlike  `lthooks-code.pdf` and similar doc files for recently added latex2e modules), hence all references to documentation entries fail.

A trivial solution is to simply stop typesetting `l3draw-code.pdf`, by removing `l3draw-code.tex` from l3build config. This means revert commit c2ef828f (Add code typesetting for l3draw [ci skip], 2018-02-19), partially or totally.
https://github.com/latex3/latex3/blob/b4f39c0316ebbc029a586bcafc0d6dfb22ac6dbd/l3experimental/l3draw/build.lua#L13

Apparently I'm not familiar with `l3doc` usages, since some simple tweaks I tried failed to add documentation part to `l3draw-code.{tex,pdf}`.

**Comment by mbertucci47** [comment]
I am in favor of the documentation and implementation being in the same file, though I can see why someone would disagree. For e.g. lthooks, there is `lthooks-doc` which only has the documentation and `lthooks-code` which has both the documentation and implementation typeset. It would be nice if `l3draw`, `l3galley`, `l3graphics`, etc. could do something similar.

**Comment by josephwright** [comment]
The aim of having the `-code` file here was to look at having a cleaner split for docs: one for the API, one for the implementation. Most of the time the two really are not needed together, hence the setup. (I do something similar in `siunitx`.)

**Comment by muzimuzhi** [comment]
Looks like this use case calls for a new class option that (globally) defines `\__codedoc_print_documented:` either as current or the same as `\prg_do_nothing:`.

**Comment by josephwright** [comment]
Well it is to some extent this was intended as an experiment in how best to document the code - there is work to do in developing best practice. So I'm interested to hear views: @mbertucci47 could you enlarge on why you favour the API and code docs in one file?

**Comment by FrankMittelbach** [comment]
@josephwright yes, but we disagree there a bit I guess. Personally I also favor the combined versions: when I'm looking at the code, I usually also look at what is said in the docs and I don't like to have two files in that case (doesn't allow me to search for example). But I also like to have the docs alone, hence my split in "doc" and "code" (which contains both doc + code).

I agree that when it comes to something like source2e it would be even nicer to have 3 versions, i.e. one where you reorder and combine the doc and the code parts, but that is a somewhat different question.

**Comment by car222222** [comment]
+1 for the @FrankMittelbach scheme to always have at least the "code+doc" file.

**Comment by mbertucci47** [comment]
@josephwright I would echo @FrankMittelbach's comments. To add a concrete example, recently I was looking through the ltproperties doc+code and wanted to know if/how `\RefProperty` expands its arguments. This type of info may be mentioned in the docs, but if you know what you're looking for it's easier to just copy "RefProperty" and search the pdf for the definition, see that it's `\property_ref:ee(n)` and know how the expansion works

**Comment by car222222** [comment]
@mbertucci47 A good example, thanks!

**Comment by car222222** [comment]
A further request: except when this is not feasible, please provide .pdf versions of *all* .dtx, etc., files (including those that are "under development").

**Comment by josephwright** [comment]
There seems to be a consensus so I will make a change: as I said, one reason for the difference here was to allow exploration of what works for people.

**Comment by josephwright** [comment]
> A further request: except when this is not feasible, please provide .pdf versions of _all_ .dtx, etc., files (including those that are "under development").

Not sure what you mean here.

**Comment by muzimuzhi** [comment]
> This type of info may be mentioned in the docs, but if you know what you're looking for it's easier to just copy "RefProperty" and search the pdf for the definition, see that it's `\property_ref:ee(n)` and know how the expansion works

@mbertucci47 A bit off-topic. [`texdef`](https://ctan.org/pkg/texdef) may also help here, with some knowledge about internal macros of `ltcmd/xparse` commands.

```
latexdef -s \RefProperty
Source code definition of 'RefProperty' could not be found.
\RefProperty is defined by (La)TeX.

\RefProperty:
macro:->\__cmd_start_expandable:nNNNNn {omm}\RefProperty  \RefProperty  \RefProperty code ?{\__cmd_expandable_grab_D:w \RefProperty (arg 1) []\__cmd_expandable_grab_m:w \__cmd_expandable_grab_m:w }

latexdef -s 'RefProperty code'
Source code definition of 'RefProperty code' could not be found.
\RefProperty code is defined by (La)TeX.

\RefProperty code:
\long macro:#1#2#3->\IfNoValueTF {#1}{\property_ref:ee {#2}{#3}}{\property_ref:een {#2}{#3}{#1}}
```

**Comment by josephwright** [comment]
I wonder how we want to name the outputs here (and more widely). For `l3kernel`, we have `interface3` and `source3`, so an API doc and an API + code doc would fit. That might suggest `l3XXX` and `l3XXX-code`?

**Comment by josephwright** [comment]
[Aside: for a document-level package like `siunitx`, `XXX` is the user docs, so `XXX-api` and `XXX-code`?]

**Comment by josephwright** [comment]
On `l3draw`, there's quite a bit of setup in the 'user' `.dtx`: how do we want to make this available to the `-code` file? Split it out, copy-paste, something else?

**Comment by FrankMittelbach** [comment]
> [Aside: for a document-level package like `siunitx`, `XXX` is the user docs, so `XXX-api` and `XXX-code`?]

I would stick with  either just XXX.pdf  or XXX-doc.pdf for the version containing just the documentation, because that is what is the comon setup and what is recognized by `texdoc` for example.  Not sure what `-api`should contain in addition or differently, can you explain? If it is just another name for `-doc`. I would not invent a new convention.

**Comment by josephwright** [comment]
> > [Aside: for a document-level package like `siunitx`, `XXX` is the user docs, so `XXX-api` and `XXX-code`?]
> 
> I would stick with either just XXX.pdf or XXX-doc.pdf for the version containing just the documentation, because that is what is the comon setup and what is recognized by `texdoc` for example. Not sure what `-api`should contain in addition or differently, can you explain? If it is just another name for `-doc`. I would not invent a new convention.

Three PDFs:

- User-level documentation
- Code-level documentation
- Code-level documentation plus typeset implementation

Currently `siunitx` doesn't have the last one

**Comment by car222222** [comment]
My point is that, wherever this is at all possible, "everything" should (starting way back at its inception as an idea or experiment) provide the last of these three, containing everything relevant to the idea:

* User-level documentation
* Code-level documentation
* Code-level documentation plus typeset implementation.

Also that, in order to make sense, this last often needs to contain "user-level" information and "api information" and etc., etc.  That is to say, all that is necessary for a complete understanding of the both the code and its various usages. 

Thus siuntix (amongst others) is seriously deficient:
> Currently siunitx doesn't have the last one

**Comment by josephwright** [comment]
> Also that, in order to make sense, this last often needs to contain "user-level" information and "api information" and etc., etc. That is to say, all that is necessary for a complete understanding of the both the code and its various usages.


I guess I tend to see it differently: the user and API docs should be self-contained, with  the code only needed for those working on it. (So personally I tend to think typesetting the code is a bit odd, as to work on it one needs it open in an editor as a `.dtx`.)

**Comment by josephwright** [comment]
Of course, I will sort whatever people feel is best practice ...

**Comment by Skillmon** [comment]
Even while working on the code, it sometimes helps to see the typeset results of the documentation around it to better understand it. Some explanations around the code are quite markup heavy, and while I can read it in my editor, ignoring the markup and focusing on the contents is indeed easier with the typeset results (especially with sections having verbatim material and markup stuff intertwined).

**Comment by car222222** [comment]
@josephwright 

A:  No one is against having any such self-contained documents.  Provided these are *additional* to the comprehensive one, provided for those who wish to view everything in one place. 

B:  There are very many reasons for looking at the documented code (especially when it is comprehensively and lucidly explained:-) that go well beyond fixing it!

**Comment by josephwright** [comment]
Back with the issue itself, the one `??` I can find in `l3backend` is where we create `color.sc`, which is documented for use in the PostScript workflow but otherwise undefined. This one is tricky: as it's backend-specific it can only be created here, but it's not really sensible to include all of `l3color` in the `l3backend` docs just to suppress this one link. I'll do that manually.

**Comment by mbertucci47** [comment]
@josephwright I'm not looking at the dev version but in my l3backend-code.pdf from texlive, I've got a dozen or more ??'s. For example `\l_graphics_search_ext_seq`, `\@makecol@hook`, etc.

**Comment by josephwright** [comment]
> @josephwright I'm not looking at the dev version but in my l3backend-code.pdf from texlive, I've got a dozen or more ??'s. For example `\l_graphics_search_ext_seq`, `\@makecol@hook`, etc.

True.  These are again tricky: we are (re)setting data, etc., so once again I'm not really sure what to do.

**Comment by FrankMittelbach** [comment]
> > @josephwright I'm not looking at the dev version but in my l3backend-code.pdf from texlive, I've got a dozen or more ??'s. For example `\l_graphics_search_ext_seq`, `\@makecol@hook`, etc.
> 
> True. These are again tricky: we are (re)setting data, etc., so once again I'm not really sure what to do.

if they are documented elsewhere, you could just use the new key and suppress the documentation. Or, alternatively, we could consider to have yet another key to say that something is documented in a different file (but I'm not that sure we want to maintain such a linkage).

**Comment by mbertucci47** [comment]
In the case of l3backend, I must admit I'm confused why there are page references at all. There is no associated user/API documentation, so shouldn't all commands be in the `no-user-doc` category?

Just for reference, here's the list of commands in l3backend giving undefined page references.
- `\__color_backend_select_cmyk:n` and others
- `\l_graphics_search_ext_seq` (four times)
- `\__pdf_backend_link_begin_goto:nnw` and others
- `\@makecol@hook`
- `color.sc`
- `TeXcolorseparation` and `separation`
- `pdf.globaldict`
- `pdf.cvs` and others
- `pdf.linkmargin`, `pdf.linkdp.pad`, and `pdf.linkht.pad`
- `pdf.rect` and others
- `pdf.dest.anchor` and others
- `pdf.bordertracking` and others
- `pdf.breaklink` and others
- `pdf.pdfmark` and others

**Comment by josephwright** [comment]
> In the case of l3backend, I must admit I'm confused why there are page references at all. There is no associated user/API documentation, so shouldn't all commands be in the `no-user-doc` category?

Well that's the question: they _do_ have user docs, but not in this particular place. (Contrast emulation of primitives, where we really mean it with 'no documentation'.)

I will pick them off and see how it looks.

**Comment by josephwright** [comment]
I think `l3backend` is now sorted - I could do with thoughts on `l3draw`.
